transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy

```diff
@@ -887,22 +873,43 @@ class MultidimensionalDataResourceService implements MultiDimensionalDataResourc
         assaysByType.keySet()?.collect {it.dataTypeName}
     }
 
-    private ArrayList<AssayConstraint> getOldAssayConstraint(Constraint assayConstraint, User user, String type) {
-        List<ObservationFact> observations = highDimObservationList(assayConstraint, user,
-                type == 'autodetect' ? defaultHDModifierCriterion : HDModifierCriterionForType(type))
+    private List<Study> selectStudiesWithDimensionSupport(Iterable<Study> studies, DimensionImpl dimension) {
+        DetachedCriteria studiesCriteria = DetachedCriteria.forClass(Study)
+            .createAlias('dimensionDescriptions', 'd')
+            .add(Restrictions.in('id', studies*.id))
+            .add(Restrictions.in('d.name', dimension.name))
 
-        List assayIds = []
-        for(def o : observations) {
-            if(o.numberValue == null) throw new DataInconsistencyException("Observation row(s) found that miss the assayId")
-            assayIds.add(o.numberValue.toLong())
+        getList(studiesCriteria)
+    }
+
+    private List<AssayConstraint> getOldAssayConstraint(Constraint assayConstraint, User user, String type) {
+        assert user instanceof DbUser
+
+        Collection<Study> userStudies = accessControlChecks.getDimensionStudiesForUser(user)
+        List<Study> assaySupportStudies = selectStudiesWithDimensionSupport(userStudies, ASSAY)
+        if (!assaySupportStudies) {
+            log.debug("No studies with assay dimension for user ${user.nam} were found.")
+            return Collections.emptyList()
         }
 
-        if (assayIds.empty){
-            return []
+        QueryBuilder builder = HibernateCriteriaQueryBuilder.forStudies(assaySupportStudies)
+
+        Criterion modifierCriterion = type == 'autodetect' ?
+                defaultHDModifierCriterion : HDModifierCriterionForType(type)
+
+        DetachedCriteria assayIdsCriteria = builder.buildCriteria(assayConstraint, modifierCriterion)
+                .setProjection(Projections.property('numberValue'))
+
+        List<BigDecimal> assayIds = getList(assayIdsCriteria)
+
+        if (assayIds.empty) {
+            Collections.emptyList()
```

Comment:
This returns an immutable list, which is fine in cases where it is currently used, but may result in unexpected runtime errors, when used somewhere else.